module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -215,6 +217,111 @@ func TestInvalidateEntity(t *testing.T) {
+			probabilityOfAdding: 0.8,
+		},
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        RandomEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+	} {
+		t.Run(fmt.Sprintf("%d-limit-%d-entities", tc.limit, tc.entityCount), func(t *testing.T) {
+			testAddRemoveEntities(t, tc.limit, tc.entityCount, tc.ejectionMode, tc.numberOfOperations, tc.probabilityOfAdding)
+		})
+	}
+}
+
+// Indexes contains an index in the pool of an inserted entity and an index of this entity
+// in the array of entities created by the EntityListFixture.
+type Indexes struct {
+	indexInPool        EIndex
+	indexInEntitiesArr int
+}
+
+func testAddRemoveEntities(t *testing.T, limit uint32, entityCount uint32, ejectionMode EjectionMode, numberOfOperations int, probabilityOfAdding float32) {
+	// create and log the seed.
+	testSeed := time.Now().UnixNano()
+	rand.Seed(testSeed)
+	fmt.Println("Seed used for this testAddRemoveEntities execution", testSeed)
+
+	pool := NewHeroPool(limit, ejectionMode)
+
+	entities := unittest.EntityListFixture(uint(entityCount))
+
+	// an array of random owner Ids.
+	ownerIds := make([]uint64, entityCount, entityCount)
+	// generate ownerId to index in the entities arrray.
+	for i := 0; i < int(entityCount); i++ {
+		ownerIds[i] = rand.Uint64()
+	}
+	// this map maintains entities currently stored in the pool.
+	var addedEntities map[flow.Identifier]Indexes
+	addedEntities = make(map[flow.Identifier]Indexes)
+
+	for i := 0; i < numberOfOperations; i++ {
+		// choose between Add and Remove with a probability of probabilityOfAdding and 1-probabilityOfAdding respectively.
+		if rand.Float32() < probabilityOfAdding || len(addedEntities) == 0 {
+			// adding an entity
+			entityToAdd := rand.Intn(int(entityCount))
+			// check that entity is not already inserted.
+			_, found := addedEntities[entities[entityToAdd].ID()]
+			if !found {
+				indexInThePool, _, ejectedEntity := pool.Add(entities[entityToAdd].ID(), entities[entityToAdd], uint64(ownerIds[entityToAdd]))
+				if indexInThePool != InvalidIndex {
+					addedEntities[entities[entityToAdd].ID()] = Indexes{indexInThePool, entityToAdd}
+				}
+				if ejectedEntity != nil {
+					delete(addedEntities, ejectedEntity.ID())
+				}
+			}
+
+		} else {
+			// randomly select an index of an entity to remove.
+			entityToRemove := rand.Intn(len(addedEntities))
+			i := 0
+			var indexInPoolToRemove EIndex = 0
+			var indexInEntitiesArray EIndex = 0
+			for _, v := range addedEntities {
+				if i == entityToRemove {
+					indexInPoolToRemove = v.indexInPool
+					indexInEntitiesArray = EIndex(v.indexInEntitiesArr)
+					break
+				}
+				i++
+			}
+
+			// remove the selected entity from the pool.
+			removedEntity := pool.Remove(indexInPoolToRemove)
+			require.Equal(t, entities[indexInEntitiesArray].ID(), removedEntity.ID(), "Removed wrong entity")
+			delete(addedEntities, entities[indexInEntitiesArray].ID())
+		}
+	}
+}
```

Comment:
At the end, once all operations are done, the test must check that all `addedEntities` are still in the `pool`, and the `pool` size matches the length of `addedEntites`. Also, the individual entities in `addedEntities` match the corresponding one that is retrieved by the `pool.Get`, similar to this:  https://github.com/onflow/flow-go/blob/master/module/mempool/herocache/backdata/heropool/pool_test.go#L618-L622

Reply:
yeah, somehow I deleted this check. I had it in the previous version. Will restore.